component/component.go

```diff
@@ -40,4 +51,23 @@ type Host interface {
 	// Context returns a context provided by the host to be used on the component
 	// operations.
 	Context() context.Context
+
+	// GetFactory of the specified kind. Returns the factory for a component type.
+	// This allows components to create other components. For example:
+	//   func (r MyReceiver) Start(host component.Host) error {
+	//     apacheFactory := host.GetFactory(KindReceiver,"apache").(receiver.Factory)
+	//     receiver, err := apacheFactory.CreateMetricsReceiver(...)
+	//     ...
+	//   }
+	// GetFactory can be called by the component anytime after Start() begins and
+	// until Shutdown() is called. Note that the component is responsible for destroying
+	// other components that it creates.
+	GetFactory(kind Kind, componentType string) Factory
+
```

Comment:
Do we need this generic method?

Reply:
This is showing 2 different approach. We either need the generic one or the 4 others.

The 4 methods are the approach that you asked for, but I am not sure it is a good idea. It requires more changes and gains are unclear. The composition of interfaces looks confusing to me.

See commit 1 vs commit 1+2 in this PR.